### PR TITLE
[refactor](jdbc catalog) Refactored JDBC Client into Abstract Base Class and Database-specific Subclasses

### PR DESCRIPTION
**Pull request by zy-kkk**
# Proposed changes

Issue Number: close #xxx

## Problem summary

This PR does the following:

1. This PR is a substantial refactor of the JDBC client architecture. The previous monolithic JDBC client has been refactored into an abstract base class `JdbcClient`, and a set of database-specific subclasses (e.g., `JdbcMySQLClient`, `JdbcOracleClient`, etc.), and the JdbcClient required config, abstract into an object. This allows for improved modularity, easier addition of support for new databases, and cleaner, more maintainable code. This change is backward-compatible and does not affect existing functionality.
2. As a result of client refactoring, OceanBaseClient can automatically recognize the mode of operation as MySQL or Oracle, so we cancel the oceanbase_mode property in the Jdbc Catalog, but due to the cancellation of the property, When creating a single OceanBase Jdbc Table, the table type needs to be filled in as oceanbase(mysql mode) or oceanbase_oracle(oracle_mode). The above work is a change in the usage behavior, please note.
3. For the PostgreSQL Jdbc Catalog, I did two things:

      1.   The adaptation to MATERIALIZED VIEW and FOREIGN TABLE is added
      2.   Fixed reading jsonb, which had been incorrectly changed to json in a previous PR

4. fix some jdbc catalog test case
5. modify oceanbase jdbc doc

And,Thanks @wolfboys for the guidance

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcMySQLClient.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        } finally {
+            close(conn);
+        }
+        return databaseNames;
+    }
+
+    @Override
+    protected void processTable(String dbName, String tableName, String[] tableTypes,
+                                Consumer<ResultSet> resultSetConsumer) {
+        Connection conn = getConnection();
```
Comment (wolfboys) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcMySQLClient.java:59]: Connection conn = null;

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcMySQLClient.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        ResultSet rs = null;
+        try {
+            DatabaseMetaData databaseMetaData = conn.getMetaData();
```
Comment (wolfboys) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcMySQLClient.java:62]: before 62 line:
`conn = getConnection();`

**Review by wolfboys — COMMENTED** [review]
We can improve this method:
```
@Override
    protected Type jdbcTypeToDoris(JdbcFieldSchema fieldSchema) {
        String trinoType = fieldSchema.getDataTypeName();
        switch (trinoType) {
            case "integer":
                return Type.INT;
            case "bigint":
                return Type.BIGINT;
            case "smallint":
                return Type.SMALLINT;
            case "tinyint":
                return Type.TINYINT;
            case "double":
                return Type.DOUBLE;
            case "real":
                return Type.FLOAT;
            case "boolean":
                return Type.BOOLEAN;
            case "date":
                return ScalarType.createDateV2Type();
            default:
                break;
        }

         if (trinoType.startsWith("decimal")) {
            String[] split = trinoType.split("\\(");
            String[] precisionAndScale = split[1].split(",");
            int precision = Integer.parseInt(precisionAndScale[0]);
            int scale = Integer.parseInt(precisionAndScale[1].substring(0, precisionAndScale[1].length() - 1));
            return createDecimalOrStringType(precision, scale);
        }
        
        if (trinoType.startsWith("char")) {
            ScalarType charType = ScalarType.createType(PrimitiveType.CHAR);
            charType.setLength(fieldSchema.columnSize);
            return charType;
        }
        
        if (trinoType.startsWith("timestamp")) {
            // timestamp with picoseconds precision, will lose precision
            return ScalarType.createDatetimeV2Type(JDBC_DATETIME_SCALE);
        }  
        
        if (trinoType.startsWith("array")) {
            String trinoArrType = trinoType.substring(6, trinoType.length() - 1);
            fieldSchema.setDataTypeName(trinoArrType);
            Type type = jdbcTypeToDoris(fieldSchema);
            return ArrayType.create(type, true);
        } 
        
        if (trinoType.startsWith("varchar")) {
            return ScalarType.createStringType();
        }

        return Type.UNSUPPORTED;
    }
```

**Comment by zy-kkk** [comment]
run buildall

**Comment by zy-kkk** [comment]
run buildall

**Comment by zy-kkk** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            case "VARCHAR2":
+            case "NVARCHAR2":
+            case "CHAR":
+            case "NCHAR":
+            case "LONG":
+            case "RAW":
+            case "LONG RAW":
+            case "INTERVAL":
+                return ScalarType.createStringType();
+            case "BLOB":
```
Comment (wolfboys) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java:114]: hi , Are the following types not supported?

```         case "BLOB":
            case "CLOB":
            case "NCLOB":
            case "BFILE":
            case "BINARY_FLOAT":
            case "BINARY_DOUBLE":
```

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        String oracleType = fieldSchema.getDataTypeName();
+        if (oracleType.startsWith("INTERVAL")) {
+            oracleType = oracleType.substring(0, 8);
+        } else if (oracleType.startsWith("TIMESTAMP")) {
+            if (oracleType.equals("TIMESTAMPTZ") || oracleType.equals("TIMESTAMPLTZ")) {
+                return Type.UNSUPPORTED;
+            }
+            // oracle can support nanosecond, will lose precision
+            return ScalarType.createDatetimeV2Type(JDBC_DATETIME_SCALE);
+        }
```
Comment (wolfboys) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java:55]: String oracleType = fieldSchema.getDataTypeName();

     if (oracleType.startsWith("TIMESTAMP")) {
            if (oracleType.equals("TIMESTAMPTZ") || oracleType.equals("TIMESTAMPLTZ")) {
                return Type.UNSUPPORTED;
            }
            // oracle can support nanosecond, will lose precision
            return ScalarType.createDatetimeV2Type(JDBC_DATETIME_SCALE);
        }

        if (oracleType.startsWith("INTERVAL")) {
            oracleType = oracleType.substring(0, 8);
        }

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+             *    In this case, Doris will use DECIMAL(s,s)
+             * 4. if we don't specify p and s for NUMBER(p,s), just NUMBER, the p and s of NUMBER are uncertain.
+             *    In this case, doris can not determine p and s, so doris can not determine data type.
+             */
+            case "NUMBER":
+                int precision = fieldSchema.getColumnSize();
+                int scale = fieldSchema.getDecimalDigits();
+                if (scale <= 0) {
+                    precision -= scale;
+                    if (precision < 3) {
```
Comment (wolfboys) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java:80]: We don't need else if:
```
if (precision < 3) {
    return Type.TINYINT;
}
if (precision < 5) {
    return Type.SMALLINT;
}
if (precision < 10) {
    return Type.INT;
}
if (precision < 19) {
    return Type.BIGINT;
}
if (precision < 39) {
// LARGEINT supports up to 38 numbers.  
    return Type.LARGEINT;
}
return ScalarType.createStringType();
```

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java (hunk 3 above)
Comment (zy-kkk) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java:114]: yes not support

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java (hunk 5 above)
Comment (zy-kkk) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcOracleClient.java:80]: dito

**Comment by zy-kkk** [comment]
run buildall

**Comment by zy-kkk** [comment]
run buildall

**Comment by zy-kkk** [comment]
run buildall

**Review by wolfboys — APPROVED** [review]
LGTM, non-binding

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**